### PR TITLE
🪄 New hcom syntax with inference

### PR DESCRIPTION
**Pull request by ralsei**
[The title of this PR is pretty bad... I'm not sure exactly what to call this change still!]

This PR adds a new syntax for `hcom`, in which the type and cofibration are read off the goal type and its boundary (so, when the goal is a `sub`). A demo of this used wherever possible is given in the provided test file `bruno.cooltt`, a formalization of much of [Cubical informal type theory: the higher groupoid structure by Bruno Bentzen](https://arxiv.org/abs/1806.08490).

The motivation for this is that leading into the graphical editor project, we want the user to have to enter as little information as possible when the computer writes an `hcom` for them.

This is currently very WIP. Namely, as discussed, the tactical `hcom_chk` needs to fix the "local soundness" property mentioned by Jon to qualify as a good `Chk` tactic. In addition, the near-identical `hfill` syntax is presently nonfunctional.

### CODE REVIEW

**Comment by ralsei** [comment]
Current things that need to be fixed:

- [ ] This would be a lot easier if `hcom` was a `Chk` tactic. We can do this by dropping a hole when `hcom` doesn't satisfy the boundary, and then removing the hole :sparkles: magically :sparkles: when the boundary is satisfied. I think this UX is generally better, and should probably apply for other similar `Syn` tactics like `coe`.
- [ ] `HFillInfer` complains about needing a true cofibration. I am not sure why.
- [ ] Sometimes, `HComInfer` infers the wrong cofibration (in the case of `bruno.cooltt`, `∂ i` instead of `∂ i ∨ ∂ k`. This is probably because the subtype is not the only source of boundary necessary. I am not sure where to get the `∂ k` from.
- [ ] This might still break the local soundness invariant mentioned. I don't fully understand the theory, so I'm relying on external insight for that.

**Comment by ralsei** [comment]
I think that the `hcom` et al as a `Chk` tactic can be worked with later, so I'm putting this up for review.

**Comment by jonsterling** [comment]
I will do my best to review this today.

File: src/frontend/Tactics.ml
Comment (jonsterling) [inline src/frontend/Tactics.ml:390]: The intention of the distinction between the `Refiner` and `Tactic` modules is that the former hosts trusted rules (ones that manually produce terms) and the latter hosts only things that are built up by combining existing trusted rules. As such I think there are two options:

1. The `hcom_chk` can be moved to the refiner.
2. The `hcom_chk` could be (potentially?) rewritten as an actual tactic, meaning something that doesn't produce terms itself but instead does so by calling the existing refinement rules. 

I'm not sure if the latter is possible, so I leave it to your judgment.

**Review by jonsterling — CHANGES_REQUESTED** [review]
I think I have got this branch up to date with `main`, though I might have made a mistake (at least the tests work).

I am basically satisfied with the functionality here; nice job on the new hcom rule. 

I left only one piece of feedback that would need to be addressed before merging --- as the new hcom code is written as a trusted rule rather than a tactic, I think it should be moved to the refiner. (Or it could be rewritten as a tactic, but that might be harder.)

File: src/frontend/Tactics.ml
Comment (ralsei) [reply src/frontend/Tactics.ml]: Done

**Review by jonsterling — APPROVED** [review]

**Pull request merged**
